### PR TITLE
NAKIJIN-304. An endpoint to remove localities. 

### PR DESCRIPTION
**Pull request by oleg1std**
Add an endpoint to delete wrong localities with the redefinition of the relations.

### CODE REVIEW

File: HappyTravel.Nakijin.Api/Services/LocationManagementService.cs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        public LocationManagementService(NakijinContext context, LocationChangePublisher locationChangePublisher)
+        {
+            _context = context;
+            _locationChangePublisher = locationChangePublisher;
+        }
+        
+        
+        // Substitutes the locality htIdToRemove for accommodations with substitutionalHtId.
+        // Zone for accommodations will be substituted if substitutionalZoneHtId is specified 
+        public async Task<Result> RemoveLocality(string removableHtId, string substitutionalHtId, string? substitutionalZoneHtId = null, CancellationToken cancellationToken = default)
```
Comment (kirillta) [inline HappyTravel.Nakijin.Api/Services/LocationManagementService.cs:27]: `substitutionalZoneHtId` isn't nullable at controller, but here it is

File: HappyTravel.Nakijin.Api/Services/LocationManagementService.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        {
```
Comment (kirillta) [inline HappyTravel.Nakijin.Api/Services/LocationManagementService.cs:28]: Could you split this method by two? It's too long

File: HappyTravel.Nakijin.Api/Services/LocationManagementService.cs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            var localitiesIds = localities.Select(a => a.Id).ToList();
+            var zoneIds = zones.Select(z => z.Id).ToList();
+            var publishTask =_locationChangePublisher.PublishRemovedLocalities(localitiesIds).AsTask();
+            await publishTask;
+            if (!publishTask.IsCompletedSuccessfully)
+                return Result.Failure($"Failed to publish locations removal. Ids '{string.Join($", ", localitiesIds)}'");
+
+            publishTask = _locationChangePublisher.PublishRemovedLocalityZones(zoneIds).AsTask();
+            await publishTask;
+            if (!publishTask.IsCompletedSuccessfully)
```
Comment (kirillta) [inline HappyTravel.Nakijin.Api/Services/LocationManagementService.cs:142]: That logic seems complex to understand. Do we have another way to achieve same results?

File: HappyTravel.Nakijin.Api/Services/LocationManagementService.cs (hunk 1 above)
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Services/LocationManagementService.cs:27]: fixed

File: HappyTravel.Nakijin.Api/Services/LocationManagementService.cs (hunk 3 above)
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Services/LocationManagementService.cs:142]: Has been changed

File: HappyTravel.Nakijin.Api/Services/LocationManagementService.cs (hunk 2 above)
Comment (oleg1std) [reply HappyTravel.Nakijin.Api/Services/LocationManagementService.cs:28]: fixed

File: HappyTravel.Nakijin.Api/Services/LocationManagementService.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        private async Task<Result> PublishLocations(List<Locality?> localities, List<LocalityZone> zones)
+        {
+            var localitiesIds = localities.Select(a => a.Id).ToList();
+            var zoneIds = zones.Select(z => z.Id).ToList();
+            var publishTask =_locationChangePublisher.PublishRemovedLocalities(localitiesIds).AsTask();
+            await publishTask;
+            if (!publishTask.IsCompletedSuccessfully)
+                return Result.Failure($"Failed to publish locations removal. Ids '{string.Join($", ", localitiesIds)}'");
+
+            publishTask = _locationChangePublisher.PublishRemovedLocalityZones(zoneIds).AsTask();
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/LocationManagementService.cs]: As I remember we decided to not publish information about zones

File: HappyTravel.Nakijin.Api/Services/LocationManagementService.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            {
+                accommodation.LocalityId = locality.Id;
+                accommodation.Modified = modified;
+                accommodation.LocalityZoneId = zone?.Id;
+                accommodation.IsCalculated = false;
+            }
+        }
+
+        
+        private async Task<Result> PublishLocations(List<Locality?> localities, List<LocalityZone> zones)
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/LocationManagementService.cs:158]: As I understand we publishinc about locations remove, but we must also publish about changes of accommodations

File: HappyTravel.Nakijin.Api/Services/LocationManagementService.cs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+                if (substitutionalZone == null)
+                    return Result.Failure($"Failed to find substitutional zone '{substitutionalZoneHtId}'");
+                    
+                if (substitutionalLocality.Id != substitutionalZone.LocalityId)
+                    return Result.Failure($"Substitutional zone '{substitutionalZoneHtId}' doesn't belong to substitutional locality {substitutionalHtId}");
+            }
+
+            var zonesToRemove = await GetZones(localityToRemove!, cancellationToken);
+            
+            var accommodationsOfLocationToRemove = await GetAccommodations(localityToRemove!, cancellationToken);
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Services/LocationManagementService.cs]: in LocationMapper file we have method ChangeLocalityDependencies, I think we can use that method, or fill this method with new logic.

File: HappyTravel.Nakijin.Api/Controllers/LocationsManagementController.cs [hunk 7]
```diff
@@ -57,11 +60,34 @@ public IActionResult MapLocations([FromBody] List<Suppliers> suppliers)
...
+        /// </summary>
+        /// <param name="removableHtId">Location id to remove </param>
+        /// <param name="substitutionalHtId">Location id for substituting the removable location</param>
+        /// <param name="substitutionalZoneHtId">Id of the substituting zone for accommodations associated with the removable location</param>
+        /// <returns></returns>
+        [ProducesResponseType((int) HttpStatusCode.OK)]
+        [ProducesResponseType((int) HttpStatusCode.BadRequest)]
+        [HttpDelete("locations/locality/remove/{removableHtId}/substitute/{substitutionalHtId}")]
+        [HttpDelete("locations/locality/remove/{removableHtId}/substitute/{substitutionalHtId}/zone/{substitutionalZoneHtId}")]
+        public async Task<IActionResult> RemoveLocalityWithSubstitution([FromRoute] string removableHtId, [FromRoute] string substitutionalHtId, [FromRoute] string? substitutionalZoneHtId = null)
```
Comment (hasmikharutyunyan) [inline HappyTravel.Nakijin.Api/Controllers/LocationsManagementController.cs:76]: I think substitutionalHtId also can be null

**Pull request closed without merging**